### ISSUE TITLE
[Bug]: No idiomatic method to prevent content files from referenced projects overwriting files in main project

### ISSUE THREAD

**Issue by nbarnwell** [dotnet/msbuild#10573]
### Issue Description

I have read about `ErrorOnDuplicatePublishOutputFiles` but that only works on Publish (i.e. not build), and isn't what I want. I want to be able to specify that a file should not be brought in from the referenced project at all. I have tried adding `PrivateAssets=none` to the `ProjectReference` in the csproj file, to no avail.

### Steps to Reproduce

1. Clone https://github.com/nbarnwell/ConflictingContentFilesTest
2. Build in Visual Studio

### Expected Behavior

The appsettings file in the output folder should be the one from the main project.

### Actual Behavior

The appsettings file in the output folder is the one from the ExternalLibrary project.

### Analysis

I believe this is a change in behaviour of project references from .NET Framework? I understand bringing in content files from _packages_, but not projects. With a package, you have deliberately packaged it to contain certain files, but that step is implicit in _project_ references, so you have no control over what is brought in. There is some setting somewhere about not bringing in content files from _transitive_ project references, but that's not what I'm looking for, either.

### Versions & Configurations

MSBuild version 17.10.4+10fbfbf2e for .NET Framework
17.10.4.21802

Microsoft Visual Studio Professional 2022 (64-bit) - Current
Version 17.10.4

**Comment by nbarnwell** [comment]
I have asked that, but apparently "transitive" content is from projects referencing projects? So for ProjectA->ProjectB->ProjectC, ProjectC's content files wouldn't end up in ProjectA's build output. That's not the problem I'm having, though. This is direct project reference's content files, not transitive ones. Although, I've read a lot of similar-but-not-the-same issues now and I'm starting to get confused with the various responses, hence the desire for the MS-approved "idiomatic" way to stop this happening.

**Comment by nbarnwell** [comment]
Out of interest, I asked Github Copilot:

> How do I prevent content files from a referenced project from overwriting files with the same name in the referencing project?

And it offered me something that I edited to:

```
  <ItemGroup>
    <!-- Exclude specific content files from being copied -->
    <Content Remove="ExternalLibrary\appsettings.json" />
  </ItemGroup>

  <Target Name="CopyReferencedContent" AfterTargets="Build">
    <!-- Manually copy content files if needed, with conditions -->
    <Copy SourceFiles="appsettings.json" DestinationFolder="$(OutputPath)" SkipUnchangedFiles="true" />
  </Target>
```

That does seem to work in my sample repro project. This can't be the idiomatic solution to this, though? Surely?

**Comment by nbarnwell** [comment]
> `<Content Remove="ExternalLibrary\appsettings.json" />` looks unnecessary in the application project.

I agree, but it's AI, so it's to be taken with a pinch of salt. This is another attempt which appears to work so far (though I've had varying results with various solutions so I don't quite trust it enough to just be a "drop-in" when this happens in future):

```
  <ItemGroup>
    <Content Update="appsettings.json">
      <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
    </Content>
    <None Remove="appsettings.json" />
    <Content Include="appsettings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </Content>
  </ItemGroup>
```

**Comment by nbarnwell** [comment]
> This might be a duplicate of [dotnet/project-system#9001](https://github.com/dotnet/project-system/issues/9001). If so, the fix [dotnet/project-system#9454](https://github.com/dotnet/project-system/pull/9454) is included in Visual Studio 2022 version 17.11.1. AFAICT, it was not backported to LTSC 17.10 versions.

That's interesting - my Visual Studio did an update only yesterday and is now on that version. I'll do some more tests.

**Comment by nbarnwell** [comment]
> This might be a duplicate of [dotnet/project-system#9001](https://github.com/dotnet/project-system/issues/9001). If so, the fix [dotnet/project-system#9454](https://github.com/dotnet/project-system/pull/9454) is included in Visual Studio 2022 version 17.11.1. AFAICT, it was not backported to LTSC 17.10 versions.

Well, I've done several builds on VS 17.11.1 on both the original project that I couldn't share, and the example repro project, and with basic csproj values (i.e. no special targets or properties beyond what VS puts in when you set a file as Content etc), I seem to get the correct appsettings.json file now.

So it seems like my issue might possibly be fixed, though it'll take a little while to earn my trust. :)

Last question - does your understanding of the fix rule out similar issues on AzDO build agents? Thanks for your help with this. I'd been losing my mind trying to make sense of all the seemingly related github issues. :S

**Comment by AR-May** [comment]
Team triage: @GangWang01 could you please try to reproduce and investigate the issue?

**Comment by GangWang01** [comment]
With the provided steps to reproduce and VS version I could not reproduce. The appsettings.json file in the output folder was from the main project, not the referenced project.
![image](https://github.com/user-attachments/assets/f9b2b473-d049-4a1b-b90d-6378387ccea4)

https://github.com/dotnet/msbuild/pull/4931 works well and gets the referenced project's appsettings.json removed from the list of items copied to output folder.
![image](https://github.com/user-attachments/assets/04dc0498-7d09-46ff-8c50-699d464c0e74)

**Comment by GangWang01** [comment]
@KalleOlaviNiemitalo Build acceleration did copy the appsettings.json file from referenced project to the output folder of main project when doing the second build with any change to the appsettings.json file from referenced project.

**Comment by GangWang01** [comment]
> This might be a duplicate of [dotnet/project-system#9001](https://github.com/dotnet/project-system/issues/9001). If so, the fix [dotnet/project-system#9454](https://github.com/dotnet/project-system/pull/9454) is included in Visual Studio 2022 version 17.11.1. AFAICT, it was not backported to LTSC 17.10 versions.

As @KalleOlaviNiemitalo pointed out, it's duplicate with the issue of project system.

**Comment by GangWang01** [comment]
Closed as it's an issue of project system and it should be fixed in 17.11 according to the milestone in the pull request https://github.com/dotnet/project-system/pull/9454.
